### PR TITLE
add v3 store classes (zarr v3 support part 1)

### PR DESCRIPTION
**Pull request by grlee77**
This PR contains changes to `storage.py` and `meta.py` for v3 support. I split this out from the full v3 branch and will have follow up PRs adding v3 support to the `Array` and `Group` classes as well as the higher level routines.

This PR defines StoreV3 versions of most existing Store types (no N5Store or ABSStore, currently). This will need some rounds of revision/feedback and there are a number of "TODO" comments in code where I was unsure about desired behavior of some aspect.

For the test cases, I tried to avoid a lot of the duplication by subclassing the v2 test classes, just overriding methods where it was necessary to make changes. These v3 classes are currently in a separate file, but could instead just be appended to the existing file if that is preferred. Another possible approach is to try keeping just the existing classes, but add parameterization based on the Zarr version.

For test coverage, there are a handful of functions like `rmdir` that support `StoreLike` inputs that only have test coverage via higher-level functions which are not yet included in this PR, so there will currently be some reduction in test coverage unless that code is removed or we add test cases.

[Description of PR]

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Comment by joshmoore** [comment]
@grlee77 : while you hammer out the actions fun, is there anything in particular you could use feedback on?

**Comment by grlee77** [comment]
> while you hammer out the actions fun, is there anything in particular you could use feedback on?

I will try to summarize a bit here later today

**Comment by grlee77** [comment]
Probably my biggest question regarding this PR is how to handle extended data types. I have not done this in a way consistent with the spec yet. Specifically there is a description of how `data_type` metadata should handle this description [here](https://zarr-specs.readthedocs.io/en/core-protocol-v3.0-dev/protocol/core/v3.0.html#array-metadata). The issue is that we do not have a URI ("value must be a URI that identifies the protocol extension") yet that we can point to?

The `_v3_core_type` dict in meta.yml contains all the dtypes in the core v3 specification. The other types we need for various tests are:

- complex (NumPy dtype.kind == 'c')
- DateTime (NumPy dtype.kind in 'mM')
- NumPy object arrays (NumPy dtype.kind == 'O')
- NumPy structured arrays (NumPy dtype.kind == 'V')
- unicode arrays (NumPy dtype.kind == 'U')
- bytes arrays (NumPy dtype.kind == 'S')

The unicode and byte array cases are currently not covered in our current Zarr test suite, but they are used in the Xarray test suite downstream.

Does each of these need to have its own separate extension URI that we can point to in the `data_type` metadata or can we just put an empty placeholder for these for now? Othewise, I can see this taking quite a while to resolve...

I don't think we should be introducing all of the `ZARR_V3_ALLOW_`... environment variables near the top of meta.yml in this PR and think we should probably remove those. I was using that initially in testing, but doubt we want all of those as public.






Should we work on creating these first?

**Comment by grlee77** [comment]
Actually, I see now that there are indeed already extensions for two of the above here:
https://zarr-specs.readthedocs.io/en/core-protocol-v3.0-dev/protocol/extensions.html

Let me update this PR to use `data_type` extensions properly

**Comment by grlee77** [comment]
> Sounds good. I do think those URIs are fairly key to the extension mechanism, so before we get too far we should have the most core ones well-defined as good examples.

I have not worked on any new URI text yet, but I did update `meta.yml` here to read and write `data_type` metadata with the format described in the v3 spec for extensions. Specifically, for the core dtypes, the JSON just has a string such as `>f8`, but for any extended type the entry is a dict with 3 keys: `extensions`, `type` and `fallback` as specified in [this section of the spec](https://zarr-specs.readthedocs.io/en/core-protocol-v3.0-dev/protocol/core/v3.0.html#array-metadata). The spec just states that it must be "an object containing the names extension, type and fallback" so a `dict` seemed an easy JSON-compatible way to store that.

The newly introduced `get_extended_dtype_info` function converts a numpy dtype's string representation to a corresponding dict of extension info. For complex and datetime64/timedelta64 types these point to the existing URI stubs. There are 4 other types that don't have a URL to point to at the moment 1.) object arrays, 2.) structured arrays, 3.) bytestring arrays, 4.) unicode arrays. Do you think those should be 4 separate extensions or should bytestring and unicode be combined into one?

**Comment by grlee77** [comment]
Also, are any of these cases where we should be specifying a fallback?. An array of complex128 can be read as a double-length array of type float64 for instance with real and imaginary values interleaved?

**Review by joshmoore — COMMENTED** [review]
> so a dict seemed an easy JSON-compatible way to store that.

*nods*

> should bytestring and unicode be combined into one?

I don't know the use case(s) for `bytestring`. Have you seen anything mentioned?

> are any of these cases where we should be specifying a fallback?. An array of complex128 can be read as a double-length array of type float64 for instance with real and imaginary values interleaved?

Is the fallback the extent of "extension subclassing"? If so, it would be good to have examples. So far it's always been time->long or similar. The complex -> 2 floats would certainly test the concept further.

cc: @dennisheimbigner @alimanfoo

File: zarr/meta.py [hunk 1]
```diff
@@ -1,14 +1,87 @@
...
+from numcodecs.abc import Codec
 
 from zarr.errors import MetadataError
 from zarr.util import json_dumps, json_loads
 
-from typing import cast, Union, Any, List, Mapping as MappingType
+from typing import cast, Union, Any, List, Mapping as MappingType, Optional
 
 ZARR_FORMAT = 2
+ZARR_FORMAT_v3 = 3
```
[the remark this answers is not in the document]
Comment (grlee77) [reply zarr/meta.py:15]: could be fine to add that. not sure how this is used

currently `ZARR_FORMAT_v3` is only used once to set the `ZARR_FORMAT` attribute in the `Metadata3` class so we could just remove it if there is not another use for it.

[the remark this answers is not in the document]
Comment (grlee77) [reply zarr/meta.py:15]: looks like the other ZARR_FORMAT is used mostly in test cases, but also when setting metadata in n5.py

**Comment by grlee77** [comment]
> I don't know the use case(s) for bytestring. Have you seen anything mentioned?

IIRC there is some test case in Xarray that currently tests zarr v2 with both unicode and bytestring arrays, but we don't currently have tests for either of those in this repository!

**Comment by DennisHeimbigner** [comment]
The usual case for bytestring is to hold something like an image.

**Comment by grlee77** [comment]
closing, these commits along with additional fixes and tests are incorporated in #898

**Pull request closed without merging**